doc/sphinx/Pacemaker_Development/helpers.rst

```diff
@@ -4,6 +4,64 @@ C Development Helpers
 .. index::
    single: unit testing
 
+Refactoring
+###########
+
+Pacemaker uses an optional tool called `coccinelle <https://coccinelle.gitlabpages.inria.fr/website/>`_
+to do automatic refactoring.  coccinelle is a very complicated tool that can be
+difficult to understand, and the existing documentation makes it pretty tough
+to get started.  Much of the documentation is either aimed at kernel developers
+or takes the form of grammars.
+
+However, it can apply very complex transformations across an entire source tree.
+This is useful for tasks like code refactoring, changing APIs (number or type of
+arguments, etc.), catching functions that should not be called, and changing
+existing patterns.
+
+coccinelle is driven by input scripts called `semantic patches <https://coccinelle.gitlabpages.inria.fr/website/docs/index.html>`_
+written in its own language.  These scripts bear a passing resemblance to source
+code patches.  They scripts tell coccinelle how to match and modify a piece of
+source code.  They are stored in ``devel/coccinelle`` and each script either
+contains a single source transformation or several related transformations.
+In general, we try to keep these as simple as possible.
+
+In Pacemaker development, we use a couple targets in ``devel/Makefile.am`` to
+control coccinelle.  The ``cocci`` target tries to apply each script to every
+Pacemaker source file, printing out any changes it would make to the console.
+The ``cocci-inplace`` target does the same but also makes those changes to the
+source files.  A variety of warnings might also be printed.  If you aren't working
+on a new script, these can usually be ignored.
```

Comment:
maybe something about only certain cocci files being run via the make target, the others being there as templates or for experimentation

Reply:
Added.